llvm/lib/Analysis/BasicAliasAnalysis.cpp

```diff
@@ -444,20 +448,22 @@ static LinearExpression GetLinearExpression(
         E = GetLinearExpression(Val.withValue(BOp->getOperand(0), false), DL,
                                 Depth + 1, AC, DT);
         E.Offset += RHS;
+        E.IsNUW &= NUW;
```

Comment:
Proof for add nuw associativity: https://alive2.llvm.org/ce/z/oiNosh

And counter-proof that the current NSW code here is incorrect: https://alive2.llvm.org/ce/z/Ga8-CT

We are missing a pre-condition: https://alive2.llvm.org/ce/z/wkP_yw

But this is unrelated to this patch...

Reply:
I have trouble actually coming up with a test case that currently gets miscompiled due to the NSW issue ... ideas welcome :)